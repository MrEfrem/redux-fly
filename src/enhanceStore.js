// @flow
import { combineReducers } from 'redux'
import isPlainObject from 'lodash/isPlainObject'
import { normalizeMountPath } from './utils/normalize'

/**
 * Function enhance an object of Redux store with the `registerReducers` method for
 * gradual registration of reducer at any nesting level of Redux store.
 * If reducer isn't pass, but pass preloaderState, then preloadedState would uses
 * how default state for new reducers.
 * @param {Object} createStore
 * @return {Function}
 *   @param {Object} reducer (optional)
 *   @param {Object} preloadedState (optional)
 *   @param {Function} enhancer (optional)
 *   @return {Object}
 *     {Function} registerReducers,
 *     {Function} subscribe,
 *     {Function} dispatch,
 *     {Function} getState,
 *     {Function} replaceReducer
 */
const enhanceStore = (createStore: Function) => {
  if (typeof createStore !== 'function') {
    throw new Error('Create store must be function')
  }
  return (reducer?: Object, preloadedState?: Object, enhancer?: Function) => {
    let store
    let reducers = {}
    let rawReducers = {}
    let rawReducersMap = []

    if (preloadedState && !isPlainObject(preloadedState)) {
      throw new Error('Preloaded state must be plain object')
    }

    // Create store with middleware for process batch actions
    if (reducer) {
      registerReducers(reducer)
      store = createStore(reducers, preloadedState, enhancer)
    } else {
      store = createStore(() => ({}), undefined, enhancer)
    }

    // Recreate reducers tree and replace them in store
    function recreateReducers () {
      reducers = {}
      function recreate(node) {
        if (isPlainObject(node)) {
          const newReducers = {}
          Object.keys(node).forEach(key => {
            const reducer = recreate(node[key])
            if (reducer) {
              newReducers[key] = reducer
            }
          })
          return combineReducers(newReducers)
        } else {
          return node
        }
      }
      reducers = recreate(rawReducers)
      if (store) {
        store.replaceReducer(reducers)
      }
    }

    // Wrap reducer for passed preloaded state
    function wrapperReducerPreloadedState(reducer, preloadedState) {
      return (state = preloadedState, action) => reducer(state, action)
    }

    /**
     * Add reducers in store
     * @param {Object} newReducers
     * @return {void}
     */
    function registerReducers(newReducers: Object) {
      if (!isPlainObject(newReducers) || Object.keys(newReducers).length === 0) {
        throw new Error('The reducers must be non empty object')
      }
      Object.keys(newReducers).forEach(key => {
        if (typeof newReducers[key] !== 'function') {
          throw new Error('Reducers must be functions')
        }
      })

      Object.keys(newReducers).forEach(key => {
        const normalizedKey = normalizeMountPath(key)
        rawReducersMap.forEach(key1 => {
          if ((normalizedKey.indexOf(key1) === 0 || key1.indexOf(normalizedKey) === 0) && key1 !== normalizedKey) {
            throw new Error(`Reducer mount path "${key1}" already busy`)
          }
        })
        const keys = normalizedKey.split(' ')
        let preloadedState1 = preloadedState
        const result: Object = keys.slice(0, -1).reduce((prev, next) => {
          if (typeof prev[next] === 'undefined') {
            prev[next] = {}
          }
          if (preloadedState1) {
            preloadedState1 = preloadedState1[next]
          }
          return prev[next]
        }, rawReducers)
        const lastKey = keys.slice(-1)
        if (preloadedState1 && preloadedState1[lastKey]) {
          result[lastKey] = wrapperReducerPreloadedState(newReducers[key], preloadedState1[lastKey])
        } else {
          result[lastKey] = newReducers[key]
        }
        rawReducersMap.push(normalizedKey)
      })
      recreateReducers()
    }

    return {
      ...store,
      registerReducers
    }
  }
}

export default enhanceStore
