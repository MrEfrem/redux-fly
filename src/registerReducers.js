// @flow
import React, { PropTypes } from 'react'
import storeShape from './utils/storeShape'
import enhanceStore from './enhanceStore'
import { createStore, compose } from 'redux'
import { checkMountPath } from './utils/checks'
import { normalizeMountPath } from './utils/normalize'
import { Provider } from 'react-redux'

/**
 * Function registers reducers in Redux store.
 * @param {Object | function} reducers
 * @return {Function}
 *   @param {Component} WrappedComponent
 *   @return {Component} new component
 */
export default (
  reducers: Function | Object,
) => {
  if (typeof reducers !== 'function' && typeof reducers !== 'object') {
    throw new Error('Reducers must be object or function')
  }
  return (WrappedComponent: any) =>
    class CreateReducer extends React.Component {
      static contextTypes = {
        store: process.env.NODE_ENV === 'test' ? PropTypes.object : storeShape
      }

      store: ?Object

      constructor(props: any, context: any) {
        super(props, context)

        let { store } = context
        this.store = null
        if (typeof store === 'undefined') {
          const composeEnhancers =
            process.env.NODE_ENV !== 'production' &&
            typeof window === 'object' &&
            window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ ?
              window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({}) : compose
          store = createStore(null, composeEnhancers(enhanceStore))
          this.store = store
        } else {
          if (typeof store.registerReducers !== 'function') {
            throw new Error('Redux store must be enhanced with redux-fly')
          }
        }

        const { reduxMountPath: propMountPath } = props

        if (typeof propMountPath !== 'undefined') {
          checkMountPath(propMountPath)
        }

        const _reducers = typeof reducers === 'function' ? reducers(props) : reducers

        if (typeof _reducers !== 'object') {
          throw new Error('Reducers must be object')
        }

        let _normReducers = {}
        Object.keys(_reducers).forEach(key => {
          _normReducers[normalizeMountPath(`${propMountPath || ''} ${key || ''}`)] = _reducers[key]
        })

        // Registration reducers
        store.registerReducers(_normReducers)
      }

      componentWillUnmount() {
        this.store = null
      }

      render() {
        if (this.store) {
          return (
            <Provider store={this.store}>
              <WrappedComponent {...this.props} />
            </Provider>
          )
        }
        return (
          <WrappedComponent {...this.props} />
        )
      }
    }
}
