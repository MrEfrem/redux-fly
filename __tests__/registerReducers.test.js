import registerReducers from '../src/registerReducers'
import { createStore } from 'redux'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import React, { PropTypes } from 'react'
import enhanceStore from '../src/enhanceStore'

test('Test invalid signature', () => {
  expect(registerReducers).toThrowError('Reducers must be object or function')
  expect(registerReducers.bind(this, 123)).toThrowError('Reducers must be object or function')
})

test('Test invalid reducers returned from function', () => {
  const Component = () => <div/>
  const ExtendedComponent1 = registerReducers(() => 123)(Component)

  expect(renderer.create.bind(renderer,
    <ExtendedComponent1/>
  )).toThrowError('Reducers must be object')
})

test('Test invalid prop reduxMountPath', () => {
  const Component = () => <div/>
  const ExtendedComponent1 = registerReducers({})(Component)

  expect(renderer.create.bind(renderer,
    <ExtendedComponent1 reduxMountPath={123}/>
  )).toThrowError('Mount path must be string')
})

test('Test invalid redux store', () => {
  const Component = () => <div/>
  const ExtendedComponent1 = registerReducers({})(Component)

  const store = createStore(() => ({}))
  expect(renderer.create.bind(renderer,
    <Provider store={store}>
      <ExtendedComponent1/>
    </Provider>
  )).toThrowError('Redux store must be enhanced with redux-fly')
})

test('Test reducers as function', () => {
  const Component = () => <div/>
  const store = createStore(null, enhanceStore)

  const ExtendedComponent1 = registerReducers((props) => {
    expect(props.text).toBe('My first todo')
  })(Component)

  expect(renderer.create.bind(renderer,
    <Provider store={store}>
      <ExtendedComponent1 text="My first todo"/>
    </Provider>
  )).toThrow()

  const ExtendedComponent2 = registerReducers(() => {
    return {
      ' ui   component ': () => ({
        text: 'My second todo'
      }),
      'greeting': () => ({ descr: 'Hello world!' })
    }
  })(Component)

  renderer.create(
    <Provider store={store}>
      <ExtendedComponent2/>
    </Provider>
  )

  expect(JSON.stringify(store.getState())).toBe('{\"ui\":{\"component\":{\"text\":\"My second todo\"}},\"greeting\":{\"descr\":\"Hello world!\"}}')
})

test('Test reducers as object', () => {
  const Component = () => <div/>
  const store = createStore(null, enhanceStore)

  const ExtendedComponent = registerReducers({
    'ui component': () => ({
      text: 'My second todo'
    }),
    'greeting': () => ({ descr: 'Hello world!' })
  })(Component)

  renderer.create(
    <Provider store={store}>
      <ExtendedComponent/>
    </Provider>
  )

  expect(JSON.stringify(store.getState())).toBe('{\"ui\":{\"component\":{\"text\":\"My second todo\"}},\"greeting\":{\"descr\":\"Hello world!\"}}')
})

test('Test reducers as object without provide redux store', () => {
  const Component = (undefined, { store }) => (
    <div>{JSON.stringify(store.getState())}</div>
  )
  Component.contextTypes = {
    store: PropTypes.object.isRequired
  }
  const ExtendedComponent = registerReducers({
    'ui component': () => ({
      text: 'My second todo'
    }),
    'greeting': () => ({ descr: 'Hello world!' })
  })(Component)

  const component = renderer.create(
    <ExtendedComponent/>
  )
  expect(component.toJSON()).toMatchSnapshot()
})

test('Test calculated mountPath with passed reduxMountPath', () => {
  const Component = (undefined, { store }) => (
    <div>{JSON.stringify(store.getState())}</div>
  )
  Component.contextTypes = {
    store: PropTypes.object.isRequired
  }
  const ExtendedComponent = registerReducers({
    'ui component': () => ({
      text: 'My second todo'
    }),
    'greeting': () => ({ descr: 'Hello world!' })
  })(Component)

  const component = renderer.create(
    <ExtendedComponent reduxMountPath=" my first   component" />
  )
  expect(component.toJSON()).toMatchSnapshot()
})
