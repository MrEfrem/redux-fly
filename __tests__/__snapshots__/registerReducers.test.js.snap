exports[`test Test calculated mountPath with passed reduxMountPath 1`] = `
<div>
  {\"my\":{\"first\":{\"component\":{\"ui\":{\"component\":{\"text\":\"My second todo\"}},\"greeting\":{\"descr\":\"Hello world!\"}}}}}
</div>
`;

exports[`test Test reducers as object without provide redux store 1`] = `
<div>
  {\"ui\":{\"component\":{\"text\":\"My second todo\"}},\"greeting\":{\"descr\":\"Hello world!\"}}
</div>
`;
